CASSANDRA-17002: Virtual table for gossip info

frankgh opened the pull request:
This commit adds a new virtual table that exposes the gossip information in tabular format.

The information is the same as the information presented through the `nodetool gossipinfo`
command, but the virtual table splits the version and value from `VersionedValue` into two
different columns. This is intented to help clients reading the vtable without the need of
parsing the version:value information (as it currently stands in gossipinfo).

The token value does not have a column. This is consistent with the gossipinfo output which
always renders ":<hidden>" for the Token value. Only the token_version column is available.

dcapwell commented on conf/cassandra.yaml:
```diff
@@ -1625,3 +1625,6 @@ enable_drop_compact_storage: false
 #     read_before_write_list_operations_enabled: true
+
+# The maximum number of gossip state transitions per peer to store for the gossip_state_transitions virtual table
+max_gossip_state_transitions_size: 10
```
please comment out, new features must not rely on the yaml so should be commented out to act more like documentation

dcapwell commented on src/java/org/apache/cassandra/gms/Gossiper.java:
```diff
@@ -148,6 +160,11 @@
+    /* list of the most recent N gossip state transitions.
+     * N is defined by the max_gossip_state_transitions_size configuration. */
+    @VisibleForTesting
+    final Map<InetAddressAndPort, Queue<GossipStateTransition>> gossipStateTransitionMap = new ConcurrentHashMap<>();
```
how costly is this if your cluster has 1k nodes?

frankgh replied:
for 1k nodes I estimate it is going to be low tens of megabytes. Maybe 10-20MB.

dcapwell commented on src/java/org/apache/cassandra/db/virtual/GossipStateTransitionsTable.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    GossipStateTransitionsTable(String keyspace)
+    {
+        super(TableMetadata.builder(keyspace, TABLE_NAME)
+                           .comment(TABLE_COMMENT)
+                           .kind(TableMetadata.Kind.VIRTUAL)
+                           .partitioner(new LocalPartitioner(InetAddressType.instance))
+                           .addPartitionKeyColumn(ADDRESS, InetAddressType.instance)
+                           .addClusteringColumn(PORT, Int32Type.instance)
+                           .addClusteringColumn(TIMESTAMP, TimestampType.instance)
+                           .addRegularColumn(HOSTNAME, UTF8Type.instance)
+                           .addRegularColumn(GENERATION, Int32Type.instance)
+                           .addRegularColumn(HEARTBEAT, Int32Type.instance)
+                           .addRegularColumn(STATUS, UTF8Type.instance)
+                           .addRegularColumn(LOAD, UTF8Type.instance)
+                           .addRegularColumn(SCHEMA, UTF8Type.instance)
+                           .addRegularColumn(DC, UTF8Type.instance)
+                           .addRegularColumn(RACK, UTF8Type.instance)
+                           .addRegularColumn(RELEASE_VERSION, UTF8Type.instance)
+                           .addRegularColumn(REMOVAL_COORDINATOR, UTF8Type.instance)
+                           .addRegularColumn(INTERNAL_IP, UTF8Type.instance)
+                           .addRegularColumn(RPC_ADDRESS, UTF8Type.instance)
+                           .addRegularColumn(SEVERITY, UTF8Type.instance)
+                           .addRegularColumn(NET_VERSION, UTF8Type.instance)
+                           .addRegularColumn(HOST_ID, UTF8Type.instance)
+                           .addRegularColumn(TOKENS, UTF8Type.instance)
+                           .addRegularColumn(RPC_READY, UTF8Type.instance)
+                           .addRegularColumn(INTERNAL_ADDRESS_AND_PORT, UTF8Type.instance)
+                           .addRegularColumn(NATIVE_ADDRESS_AND_PORT, UTF8Type.instance)
+                           .addRegularColumn(STATUS_WITH_PORT, UTF8Type.instance)
+                           .addRegularColumn(SSTABLE_VERSIONS, UTF8Type.instance)
+                           .build());
+    }
```
Personal preference, you can use the CREATE TABLE string syntax, calling `CreateTableStatement.parse(format(cql, table), SchemaConstants.SYSTEM_KEYSPACE_NAME)` will do that for you

frankgh replied:
Well, it involves parsing. I personally like this approach better, and it is more consistent with the vtables I have looked at.

dcapwell commented on src/java/org/apache/cassandra/db/virtual/GossipStateTransitionsTable.java:
```diff
@@ -0,0 +1,145 @@
+                           .comment(TABLE_COMMENT)
+                           .kind(TableMetadata.Kind.VIRTUAL)
+                           .partitioner(new LocalPartitioner(InetAddressType.instance))
+                           .addPartitionKeyColumn(ADDRESS, InetAddressType.instance)
+                           .addClusteringColumn(PORT, Int32Type.instance)
```
heh, `ClientsTable` does this (`(address), port`), `InternodeOutboundTable/InternodeInboundTable` does `(address, port)`, `PendingHintsTable` does `(host_id)`... we are not consistent cross tables... heh

dcapwell replied:
started a side conversation with @clohfink and he was making the point that we should really consider how this table will be used so we know what the PK should be, one argument is "last N gossip transitions makes more sense, so PK=time is best".

@frankgh what is the use case you are solving for?

frankgh replied:
One of the use cases is to provide the ability of checking the state of a host. I am thinking that we partition by `(address, port), timestamp` for this particular use case.

Another use case is debugging. For debugging it also seems important to have a view from the (host, port).

What do you think?

dcapwell commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -570,6 +570,11 @@
+    /**
+     * The maximum number of gossip state transitions per peer to store for the gossip_state_transitions virtual table
+     */
+    public int max_gossip_state_transitions_size = 10;
```
should be `gossip_state_max_transition_size` to be in-line with https://issues.apache.org/jira/browse/CASSANDRA-15234

skoppu22 replied:
Should it be gossip_state_max_transition**s**_size ? Also, would be helpful if we have 'per_peer' in the name, clearly indicating this tunable is 'per peer'.

dcapwell commented on src/java/org/apache/cassandra/gms/Gossiper.java:
```diff
@@ -393,6 +411,51 @@ public void onChange(InetAddressAndPort endpoint, ApplicationState state, Versio
             }
         });
+
+        subscribers.add(new IEndpointStateChangeSubscriber()
```
you are dropping the event type, in cases like `onRemove` this can be an internal decision, so 0 gossip state change will be present in the row, so the only difference would be the timestamp on creating the `GossipStateTransition` and not the actual contents of the state.

dcapwell commented on test/unit/org/apache/cassandra/db/virtual/GossipStateTransitionsTableTest.java:
```diff
@@ -0,0 +1,96 @@
+            {
+                assertThat(row.getColumns().size()).isEqualTo(24);
+                assertThat(row.getTimestamp("timestamp").getTime()).isGreaterThan(testStartMillis)
+                                                                   .isLessThan(currentTimeMillis());
```
this can be flaky as millis is free to go backwards (yay clocks!), and also free to not change.

frankgh replied:
I believe `nanoTime()` can go backwards; millis is guaranteed to at least not go backwards. I actually forget which one is guaranteed. But I agree that this is risky, so I will think about something for this

dcapwell replied:
`nano` can't go backwards its based off cpu cycles, `millis` can and does go backwards.  A simple way to trigger that is ntp

dcapwell commented on test/unit/org/apache/cassandra/db/virtual/GossipStateTransitionsTableTest.java:
```diff
@@ -0,0 +1,96 @@
+    @Before
+    public void config()
+    {
+        table = new GossipStateTransitionsTable(KS_NAME);
```
unsafe pattern; is there a reason you need to keep creating the table? I had to "fix"/"work around" this pattern in https://issues.apache.org/jira/browse/CASSANDRA-17295

A safer pattern is to set this up in the `setup` method, you also don't need to save the table to this class

dcapwell commented on test/unit/org/apache/cassandra/gms/GossiperTest.java:
```diff
@@ -190,6 +191,7 @@ public void testLargeGenerationJump() throws UnknownHostException, InterruptedEx
         {
             // clean up the gossip states
             Gossiper.instance.endpointStateMap.clear();
+            Gossiper.instance.gossipStateTransitionMap.clear();
```
to be more maintainable should we not have a `cleanup` method?  Seeing a lot of other maps not cleaned up, so this gets corrupted

frankgh replied:
sounds good

skoppu22 commented on test/unit/org/apache/cassandra/gms/GossiperTest.java:
```diff
@@ -414,11 +417,95 @@ public void testNotFireDuplicatedNotificationsWithUpdateContainsOldAndNewState()
+            // the list of recent gossip state transtitions includes 2 transitions from host 1 and 1 transition
+            // from host 2
+            List<GossipStateTransition> gossipStateTransitions = Gossiper.instance.recentGossipStateTransitions();
+            assertThat(gossipStateTransitions.size()).isEqualTo(3);
```
Seems like removing a endpoint from membership is not tested,  we can add below lines to test that code path as well

```
           Gossiper.instance.replacedEndpoint(remoteHostAddress1);
            List<GossipStateTransition> gossipStateTransitionsAfterReplace = Gossiper.instance.recentGossipStateTransitions();
            assertThat(gossipStateTransitionsAfterReplace.size()).isEqualTo(1);
```

skoppu22 commented on doc/modules/cassandra/pages/new/virtualtables.adoc:
```diff
@@ -290,6 +292,25 @@ SELECT total - progress AS remaining
+....
+cqlsh:system_views> SELECT * FROM gossip_state_transitions;
+
+ address   | port | timestamp                       | dc   | generation | heartbeat | host_id | hostname  | internal_address_and_port | internal_ip | load | native_address_and_port | net_version | rack | release_version | removal_coordinator | rpc_address | rpc_ready | schema | severity | sstable_versions | status | status_with_port | tokens
```
This table format has many columns, only one row. It's very difficult to map column name with corresponding value, as wrapping to multiple lines. I feel having them as rows will make viewing and also scrolling comfortable. What do you think?

skoppu22 replied:
Would like to have @yifan-c to comment as we recently created few virtual tables with metrics as rows.

frankgh replied:
The table has multiple rows, potentially thousands. I think we can query a subset of the columns for documentation purposes

skoppu22 replied:
Oh yeah, missed this

skoppu22 commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -3508,6 +3508,18 @@ public static boolean autoOptimiseIncRepairStreams()
         return conf.auto_optimise_inc_repair_streams;
     }
 
+    public static int getMaxGossipStateTransitionsSize()
```
If 'max_gossip_state_transitions_size' is renamed, I think need to rename these get and set functions to be inline with the new name.

skoppu22 commented on src/java/org/apache/cassandra/db/virtual/GossipStateTransitionsTable.java:
```diff
@@ -0,0 +1,145 @@
+ * A {@link VirtualTable} that returns the {@code max_gossip_state_transitions_size} most recent gossip state
+ * transitions. The table only considers peers in the ring (it doesn't track transitions for evicted peers).
+ *
+ * <p>The {@code max_gossip_state_transitions_size} property can be configured in the {@code cassandra.yaml}
```
Need to update this comment if we are removing param from the yaml file

skoppu22 commented on src/java/org/apache/cassandra/db/virtual/GossipStateTransitionsTable.java:
```diff
@@ -0,0 +1,145 @@
+    private static final String HEARTBEAT = "heartbeat";
+
+    // Fields coming from @org.apache.cassandra.gms.ApplicationState
+    private static final String STATUS = "status";
```
(Optional) As these names are already part of ApplicationState enum, can't we just reuse the same enum here ?

skoppu22 commented on src/java/org/apache/cassandra/db/virtual/GossipStateTransitionsTable.java:
```diff
@@ -0,0 +1,145 @@
+
+    /**
+     * @return a {@link org.apache.cassandra.db.virtual.AbstractVirtualTable.DataSet} with data populated with
+     * the {@code max_gossip_state_transitions_size} most recent gossip state transtitions
```
This comment becomes obsolete when we change the tunable name 'max_gossip_state_transitions_size'. Probably good to not mention tunable name here?

skoppu22 commented on src/java/org/apache/cassandra/gms/GossipStateTransition.java:
```diff
@@ -0,0 +1,145 @@
+
+    private String getTokensString(VersionedValue tokens)
+    {
+        return tokens == null ? null : tokens.version + ":<hidden>";
```
Just trying to understand, why are we appending "hidden" here ?

skoppu22 commented on src/java/org/apache/cassandra/gms/Gossiper.java:
```diff
@@ -501,6 +563,24 @@ public long getEndpointDowntime(InetAddressAndPort ep)
+    public List<GossipStateTransition> recentGossipStateTransitions()
+    {
+        // flatten the gossipStateTransitions map into a list sorted by the timestamp
+        return gossipStateTransitionMap.values().stream()
```
Thinking loud. As you mentioned this map size can be up to tens of megabytes, preparing a List from this can further pressure the memory with more megabytes. So instead of creating a list here, if we just return reference to the map and let the caller directly read the map and fill the vtable, can reduce memory footprint. What do you think?

frankgh committed:
CASSANDRA-17002: Virtual table for gossip info

This commit adds a new virtual table that exposes the gossip information in tabular format.

The information is the same as the information presented through the `nodetool gossipinfo`
command, but the virtual table splits the version and value from `VersionedValue` into two
different columns. This is intented to help clients reading the vtable without the need of
parsing the version:value information (as it currently stands in gossipinfo).

The token value does not have a column. This is consistent with the gossipinfo output which
always renders ":<hidden>" for the Token value. Only the token_version column is available.

frankgh committed:
higher

frankgh commented:
Closing and pushing a new branch
